{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7ce53ae2-e9af-446b-a76a-824b284911a5",
   "metadata": {},
   "source": [
    "## Load, clean, and combine TCGA-DLBC RNA-seq count files\n",
    "\n",
    "This notebook processes the raw RNA-seq gene expression count files downloaded from the Genomic Data Commons (GDC) for the TCGA-DLBC project. The goal of this step is to convert the individual STAR-generated count files into a single, tidy gene-by-sample count matrix suitable for downstream normalization and analysis.\n",
    "\n",
    "The notebook reads all STAR “Counts” TSV files from the local data directory, removes non-gene summary rows produced by the STAR workflow, and cleans gene identifiers by removing Ensembl version suffixes. For each file, unstranded gene counts are extracted and stored as a pandas Series indexed by gene ID. These series are then concatenated column-wise to form a matrix with genes as rows and tumor samples as columns.\n",
    "\n",
    "Sample-level quality control is performed by computing library sizes, defined as the total number of sequencing reads per sample. This provides a basic check on sequencing depth and ensures that all samples have adequate coverage. Column names are then cleaned and standardized by mapping file identifiers to filenames using the GDC manifest and extracting TCGA sample barcodes to produce interpretable and consistent sample labels.\n",
    "\n",
    "Finally, the cleaned gene-by-sample count matrix is saved in an efficient columnar format (Parquet) for use in subsequent normalization, exploratory analysis, predictive modeling, and inferential procedures. This step establishes a reproducible and tidy representation of the raw gene expression data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ad54f7ca-bc76-4af3-a60f-0c8aff4c50c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "48"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "tsv_paths = sorted(Path(\"../data/gdc\").rglob(\"*.tsv\"))\n",
    "len(tsv_paths)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f34c2861-1852-4020-8b1c-474da00f4b63",
   "metadata": {},
   "source": [
    "Define loader function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2f867412-b6ec-4228-b462-1ffad9a24915",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_counts_tsv(path: Path) -> pd.Series:\n",
    "    # Skip metadata lines that start with '#'\n",
    "    df = pd.read_csv(path, sep=\"\\t\", comment=\"#\")\n",
    "\n",
    "    # Drop STAR-style summary rows (they are N_* not __*)\n",
    "    df = df[~df[\"gene_id\"].astype(str).str.startswith(\"N_\")].copy()\n",
    "\n",
    "    # Clean Ensembl version suffix\n",
    "    genes = df[\"gene_id\"].astype(str).str.replace(r\"\\.\\d+$\", \"\", regex=True)\n",
    "\n",
    "    # Use unstranded gene counts\n",
    "    counts = pd.to_numeric(df[\"unstranded\"], errors=\"raise\")\n",
    "\n",
    "    return pd.Series(counts.values, index=genes.values, name=path.parent.name)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "916ae3ff-686c-4e90-b625-c8c906a41067",
   "metadata": {},
   "source": [
    "Read and combine all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ac7f662f-c152-43ce-aac0-d87a95aea3bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(60660, 48)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "series_list = [read_counts_tsv(p) for p in tsv_paths]\n",
    "counts = pd.concat(series_list, axis=1)\n",
    "\n",
    "counts.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "26a5c542-6d66-4617-8606-3619c924be70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    4.800000e+01\n",
       "mean     4.348995e+07\n",
       "std      1.058509e+07\n",
       "min      2.327589e+07\n",
       "25%      3.832623e+07\n",
       "50%      4.255281e+07\n",
       "75%      5.206553e+07\n",
       "max      6.433010e+07\n",
       "dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_sizes = counts.sum(axis=0)\n",
    "library_sizes.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d67941f0-cede-4f8a-92c1-1e3c02b73aae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['6f4e8f95-44f7-41c5-8af9-19212eff36d6.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '2fb12ae8-cb05-41bb-a567-0c7e300bb07e.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       'f7a04906-7cb5-4f46-a885-699969be5405.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '5ae63973-9111-43c9-9220-7fd7510294e1.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       'c70162ed-dafc-4a4c-b7dc-354d2c06b2c9.rna_seq.augmented_star_gene_counts.tsv'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest = pd.read_csv(\n",
    "    \"../manifests/tcga_dlbc_star_counts_manifest.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "id_to_filename = dict(zip(manifest[\"id\"], manifest[\"filename\"]))\n",
    "\n",
    "counts = counts.rename(columns=id_to_filename)\n",
    "counts.columns[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf7a53ec-7a28-4562-9f6d-ac28c8bde68c",
   "metadata": {},
   "source": [
    "Extract TCGA barcodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "08078ee5-a607-4f62-846b-5272ea04830b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['6f4e8f95-44f7-41c5-8af9-19212eff36d6.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '2fb12ae8-cb05-41bb-a567-0c7e300bb07e.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       'f7a04906-7cb5-4f46-a885-699969be5405.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '5ae63973-9111-43c9-9220-7fd7510294e1.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       'c70162ed-dafc-4a4c-b7dc-354d2c06b2c9.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '528116e8-c133-49c9-a532-18ad99a6fe52.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '70510d37-dc1a-433b-84ee-d0f7d9ca1c3d.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       'cba52f4f-eb3a-4475-8735-d5c9fd355b3d.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       'e492067e-bded-4222-a2be-302f6d39160e.rna_seq.augmented_star_gene_counts.tsv',\n",
       "       '18ba7bf2-59d3-4c59-acdc-088cf3d4e5a6.rna_seq.augmented_star_gene_counts.tsv'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_tcga_barcode(s: str) -> str:\n",
    "    m = re.search(r\"(TCGA-[A-Z0-9]{2}-[A-Z0-9]{4})\", s)\n",
    "    return m.group(1) if m else s\n",
    "\n",
    "counts = counts.rename(\n",
    "    columns={c: extract_tcga_barcode(c) for c in counts.columns}\n",
    ")\n",
    "\n",
    "counts.columns[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98a0f647-74b0-4b43-a53e-70f536cc4436",
   "metadata": {},
   "source": [
    "Save the matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "24d961fb-24aa-4137-ab12-5cc984f4f549",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts.to_parquet(\"../data/dlbc_star_counts.parquet\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
